File: leetcode/algorithms/hash-table/next-greater-element-i.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Next Greater Element I"
*/

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        HashMap<Integer,Integer> indexMap = new HashMap<Integer,Integer>();
        
        for(int i=0;i<nums2.length;++i){
            indexMap.put(nums2[i],i);
        }
        
        int[] ans = new int[nums1.length];
        int ansIndex=0;
        for(int qNum : nums1){
            int currentAns=-1;
            for(int i=indexMap.get(qNum)+1;i<nums2.length;++i){
                if(nums2[i]>qNum){
                   currentAns=nums2[i]; 
                    break;
                }
            }
            ans[ansIndex]=currentAns;
            ++ansIndex;
        }
        return ans;
        
    }
}

File: leetcode/algorithms/stack/next-greater-element-i.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Next Greater Element I"
*/

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */

const _private_stack_map = new WeakMap();

const Stack = function(){
	const _stackData = {
		stack: []
	}
	_private_stack_map.set(this,_stackData);
	
}


Stack.prototype.push = function(val){
	let stackData = _private_stack_map.get(this);
	stackData.stack.push(val);
}

Stack.prototype.pop = function(){
	let stackData = _private_stack_map.get(this);
	return stackData.stack.pop();
}

Stack.prototype.peek = function(){
	let stackData = _private_stack_map.get(this);
	return stackData.stack.length===0 ? null : stackData.stack[stackData.stack.length-1];
}

Stack.prototype.getSize = function(){
	let stackData = _private_stack_map.get(this);
	return stackData.stack.length;
}

var nextGreaterElement = function(nums1, nums2) {
    const greaterStack = new Stack();
    const numToNextGreaterMap = new Map();
    
    nums2.forEach((num,index) => {
        while(greaterStack.peek()!=null && greaterStack.peek()<num){
            numToNextGreaterMap.set(greaterStack.pop(),num);
        }
        greaterStack.push(num);
    })
    
    return nums1.reduce((arr,num) => {
        arr.push(numToNextGreaterMap.has(num) ? numToNextGreaterMap.get(num) : -1);
        return arr;
    },[])
    
    
};